### DISCUSSION TITLE
Is there a method or macro to get the band-format-size (in bytes)?

### DISCUSSION THREAD

**Discussion by VivitionDeveloper** [libvips/libvips#4133] [General]
Just to calculate the amout of memory, I need something like 'how many bytes to a pixel occupy'. In pseudo code:

```
size_t iBytes = imgWidth * imgHeight * iNumChannels * iBandFormatSize
```

The band format size depends on the VipsBandFormat enum, see https://www.libvips.org/API/current/file-format.html, but I don't know whether there is a macro that returns the information I need. I'm not sure whether e.g. VIPS_IMAGE_SIZEOF_PEL can be used. Actually, when I know 'bytes-per-pixel', then I'm fine. Does libvips provide such macro (or method)?

**Comment by VivitionDeveloper** [comment]
I have found another piece of documentation: [vips_format_sizeof](https://www.libvips.org/API/current/libvips-header.html#vips-format-sizeof), which seems to do what I need.

**Reply by jcupitt** [reply]
Yes, and there are a few macros based off that:

https://www.libvips.org/API/current/VipsImage.html#VIPS-IMAGE-SIZEOF-ELEMENT:CAPS

element, pixel, line and image, plus a matching set for regions.

**Reply by VivitionDeveloper** [reply]
Thanks and sorry for taking your time for this question.

**Comment by VivitionDeveloper** [comment]
Is there a similar macro to retrieve which band format belongs to a VipsInterpretation? I.e. how many bytes belong to the interpretations listed at [VipsInterpretatation](https://www.libvips.org/API/current/VipsImage.html#VipsInterpretation).org/API/current/VipsImage.html#VipsInterpretation).

The reason is that I'm busy with test code that can easily switch from LAB (8-bit), LABS (16-bit) and HSV (8-bit) to e.g. check the impact of colorspaces on performance (brighten, gamma correction, tonelut (applies to in-max/out-max parameters) and saturation). For now tonelut is the one that needs to know in-max/out-max based on the provided interpretation. Of course, adjusting luminance is in LAB(S) on the L-channel (band[0]) and HSV is on the V-channel (band[2]). I really like if you can give a helpful hint.

I even do not know whether tonelut can be applied to HSV images, as the documentation for [tonelut ](https://www.libvips.org/API/current/libvips-create.html#vips-tonelut).org/API/current/libvips-create.html#vips-tonelut) meanly speaks about LAB.

**Reply by jcupitt** [reply]
They are separate things --- you can have a float RGB image, for example, ie. three 32-bit floats per pixel, with 0 - 255 interpreted as black to white for each channel. This is one way of representing HDR, since you can have values well outside the usual black to white range.

So `format` is a physical thing which says what binary format the image is using for pixels (uint8, complex64 etc.), and `interpretation` is a hint saying how those numeric values should be interpreted for display to humans. 

`LAB` means the CIE numeric range with 0-100 for L\* and -100 to +100 for a\*b\*. Values outside the range are fine. The `format` can be anything, but float and double make most sense.

`LABS` is 0-32767 for L\* and -32768 to +32767 for a\*b\*. It's handy if you want a fast integer path, but a pain to use from something like python since you'll have to scale all your constants. It works best with the `format` set to int16, but again any format is fine.

Other values for `interpretation` have fairly obvious rules, the docs have the details, though they could probably be expanded, looking at them again.

https://www.libvips.org/API/current/VipsImage.html#VipsInterpretation

**Reply by jcupitt** [reply]
`tonelut` is designed for the L band of LABS images, ie. 0-32767 for black to white. You can use the lookup table on anything, of course, but you'd need to transform the image to that range first for it to be meaningful.

I wouldn't use `HSV`, it's a pretty terrible colourspace and is only there for compatibility with legacy code. The CIE equivalent is the **much** better LCh:

https://en.wikipedia.org/wiki/HCL_color_space

**Reply by jcupitt** [reply]
Sorry, I keep thinking of more things to say.

`vips_image_guess_interpretation()` is useful -- it looks at the image and checks if the `interpretation` value makes sense. For example, a one-band image that has been tagged as RGB clearly has the wrong tag. If the `interpretation` looks bad, it tries to pick a sensible alternative for you.

https://www.libvips.org/API/current/libvips-header.html#vips-image-guess-interpretation

**Reply by VivitionDeveloper** [reply]
Thanks again for excellent answers. As the real trigger for me was performance consideration, then one question left:
Does it makes much sense to use LAB or LCH instead of LABS for better performance?

If you think the performance is comparable, I'm fine and will skip LAB or LCH. Otherwise, I have to check LAB or LCH and make a choice between these, if they outperform LABS.

I was already busy with trying LAB, and vips_gamma with exponent = 1.2 behave differently for LABS (output is fine) and LAB (output becomes darker). If you are willing to give a hint, your are welcome, but it is a little out of the scope of this discussion.

Pseudo code (dAmount = 20):
```
img = img.img.colourspace(VipsInterpretation::VIPS_INTERPRETATION_LAB)
auto labs = img.bandsplit();
labs[0] = labs[0].gamma(vips::VImage::option()->set("exponent", 1 + dAmount / 100.0));
img = vips::VImage::bandjoin(labs)
	.copy(vips::VImage::option()->set("interpretation", VipsInterpretation::VIPS_INTERPRETATION_LAB));
```

**Reply by jcupitt** [reply]
`LABS` is supposed to be quicker. I tried a tiny benchmark:

```python
#!/usr/bin/env python3

import sys
import pyvips

image = pyvips.Image.new_from_file(sys.argv[1], access='sequential')
image = image.colourspace(sys.argv[3])
image = image.gaussblur(100)
image.write_to_file(sys.argv[2])
```

With a 6k x 4k test image I see:

```
$ time ./blur.py ~/pics/nina.jpg x.jpg srgb

real	0m0.763s
user	0m4.997s
sys	0m1.690s
```

That's blurring an 8-bit image, so you get the vector path and it's pretty quick. With LABS I see:

```
$ time ./blur.py ~/pics/nina.jpg x.jpg labs

real	0m4.542s
user	2m16.792s
sys	0m1.240s
```

Now that's 16-bit pixels plus rgb -> lab -> rgb conversion. It's off the vector path and a lot slower.

With 32-bit float pixels:

```
$ time ./blur.py ~/pics/nina.jpg x.jpg lab

real	0m4.817s
user	2m21.984s
sys	0m1.614s
```

Only very slightly slower, which is pretty amazing, modern CPUs have extremely quick float units.

There's a difference in memory use of course:

```
$ /usr/bin/time -f %M:%e ./blur.py ~/pics/nina.jpg x.jpg labs
996704:4.54
$ /usr/bin/time -f %M:%e ./blur.py ~/pics/nina.jpg x.jpg lab
1224664:4.74
```

This PC has a lot of cores, you'll see much lower memory use if you drop the threadpool size.

**Reply by jcupitt** [reply]
> Only very slightly slower, which is pretty amazing, modern CPUs have extremely quick float units.

I'd expect LABS to be quite a bit quicker on something lower powered, like a raspberry pi or an old laptop.

**Reply by VivitionDeveloper** [reply]
Thanks again.

**Reply by VivitionDeveloper** [reply]
The involved hardware is either a powerful Epyc server (during development) or a quite powerful AMD 6 or 8 core CPU (varying from Ryzen 7 1700 to a 5700).

**Reply by VivitionDeveloper** [reply]
> `tonelut` is designed for the L band of LABS images, ie. 0-32767 for black to white. You can use the lookup table on anything, of course, but you'd need to transform the image to that range first for it to be meaningful.
> 
> I wouldn't use `HSV`, it's a pretty terrible colourspace and is only there for compatibility with legacy code. The CIE equivalent is the **much** better LCh:
> 
> https://en.wikipedia.org/wiki/HCL_color_space

Mistake of me: I meant maplut. Does this also apply to maplut, i.e. makes it difference in using LAB (8-bit) or LCh color spaces?

**Reply by jcupitt** [reply]
I wouldn't use LAB for gamma changes. Gamma relates to the amount of light entering the camera at each point, and the transfer function that maps that to display light. 

For "amount of light" it's helpful to imagine something like photons-per-second (pps). If you don't have a numeric value in your pixel which you can relate mathematically to pps, gamma doesn't make any sense.

CIELAB is a highly non-linear colourspace, and gamma won't mean much. Why would you change the gamma of a*, for example? You need to convert to a linear space (ie. `pps = K * pixel-value`) like XYZ or scRGB, adjust the gamma, then go back to your main space.

**Reply by jcupitt** [reply]
`maplut` only works for 8- and 16-bit images, since pixel values have to be indexes. To map an LAB (for example) image, you'd need to scale the three float values down to 0-65535 (or 0-255 I guess) first.

**Reply by VivitionDeveloper** [reply]
> I wouldn't use LAB for gamma changes. Gamma relates to the amount of light entering the camera at each point, and the transfer function that maps that to display light.
> 
> For "amount of light" it's helpful to imagine something like photons-per-second (pps). If you don't have a numeric value in your pixel which you can relate mathematically to pps, gamma doesn't make any sense.
> 
> CIELAB is a highly non-linear colourspace, and gamma won't mean much. Why would you change the gamma of a*, for example? You need to convert to a linear space (ie. `pps = K * pixel-value`) like XYZ or scRGB, adjust the gamma, then go back to your main space.

I only want to change the L-channel, not A or B and then back to sRGB.

**Reply by VivitionDeveloper** [reply]
> `maplut` only works for 8- and 16-bit images, since pixel values have to be indexes. To map an LAB (for example) image, you'd need to scale the three float values down to 0-65535 (or 0-255 I guess) first.

Then this is my mistake. I thought LAB was 8-bit and LABQ float. I expected that, as in the used example in https://github.com/libvips/libvips/discussions/4036 the min/max values have been set to 255 for VIPS_INTERPRETATION_LAB.

I now get really confused and wonder the size of the values in VIPS_INTERPRETATION_LCH color space too.

**Reply by jcupitt** [reply]
> I only want to change the L-channel, not A or B and then back to sRGB.

That might look a bit odd -- most people would expect saturation to change with gamma. It would probably be best to just change gamma in sRGB.

`LCh` is 0-100 for L, 0-150 (or so, C can get a bit bigger even) for C, and h is in degrees, so 0-359.

**Reply by VivitionDeveloper** [reply]
> > I only want to change the L-channel, not A or B and then back to sRGB.
> 
> That might look a bit odd -- most people would expect saturation to change with gamma. It would probably be best to just change gamma in sRGB.
> 
> `LCh` is 0-100 for L, 0-150 (or so, C can get a bit bigger even) for C, and h is in degrees, so 0-359.

I'm getting lost a little. For adjusting saturation I do something like (all operations in LABS colorspace):
```
img = img.linear({ 1 , 1 + dAmount, 1 + dAmount }, 0);
```
Code was borrowed from https://github.com/libvips/libvips/issues/825.

For brighten, the code is:
```
labs[0] = labs[0].gamma(vips::VImage::option()->set("exponent", 1 + dAmount / 100.0));
this->img = vips::VImage::bandjoin(labs)
	.copy(vips::VImage::option()->set("interpretation", VipsInterpretation::VIPS_INTERPRETATION_LABS));
```
I don't remember where this code was borrowed from. Is this code wrong?

**Reply by VivitionDeveloper** [reply]
If have time to comment, please let me know which methods are preferred to adjust saturation and brightness. I did some code changes at my side, and add enough, adjusting brightness (using gamma) now behaves differently (images get far too bright). Not sure why. I really appreciate to get additional information.

Also: I discovered that switching to LAB or LCH, that the underlying type is float, not uchar (as expected by me, based on the set in-max/out-max in https://github.com/libvips/libvips/discussions/4036#discussioncomment-10030866). But after even reading even more documentation, I found that [it](https://www.libvips.org/API/current/libvips-colour.html#libvips-colour.description) says (although not very explicit) that LAB and LCH use float. But then, why using 255 (uchar max) as in-max/out-max. I really got confused. Do I mix things in a wrong way?

Finally: I found with performance testing (similar to what you have showed in the blur example) that LAB is 'slow' compared to sRGB calculations, but I guess there is no faster method for uniform (to human eyes) adjustments of luminance. Can you please confirm (or preferred, deny and tell about the silver bullet)?

**Reply by VivitionDeveloper** [reply]
Please, can you give me an update. My last questions did not get an answer.

**Reply by jcupitt** [reply]
Ooop, sorry.

I think it depends what style of adjust you want to make, which in turn I guess depends on what your users expect and prefer.

You can attempt to make a perceptually uniform adjustments in CIELAB. In this space you can change *L\** to get uniform brightness changes, so (for example, in python again):

```python
lab = image.colourspace("lab")
lab *= [1.1, 1, 1]
```

makes every pixel brighter by 10% of its current brightness, in "units", where a unit is a just-discerable-difference for a standard observer. Or you can change saturation with maybe:

```python
lab = image.colourspace("lab")
lab *= [1, 1.1, 1.1]
```

Which would change *a\*b\** in the same way. However, while this is perceptually correct, few users will ever have adjusted an image like this, and may not like the results.

Something like photoshop will do a "physical" rather than a perceptual change, so they will use a linear space to scale the number of "photons" in the light source:

```python
scrgb = image.colourspace("scrgb")
scrgb *= 1.1
```

This will effectively increase the amount of light illuminating the scene, so brightness and saturation will both change.

There's no physically correct way to change saturation without changing brightness, so scaling *a\*b\** would probably be your best bet. A poor second (imo) would be changing S in HSV.

Performance: you could use `labs`, but it seems to have little benefit in my tests on current x86 hardware, since they do float so quickly. You might as well stick with plain `lab` (ie. float pixels) for convenience.

You could try and use LUTs for speed, but I'd only bother if these operations start to seem expensive when you test. The obvious code above ^^^ should be pretty quick.

**Reply by VivitionDeveloper** [reply]
Many thanks for your always valuable feedback.
